### ISSUE TITLE
The state of React component testing

### ISSUE THREAD

**Issue by trevorsenior** [willowtreeapps/react-sapling#51]
#### TL;DR

Discussion around React component testing that will influence the default way sapling tests component projects. Share what worked & what didn't work when testing React components even if it's outside the realm of `react-sapling`.

#### Alright, so

We currently use Karma as our test runner. This was done to load in any arbitrary webpack configuration for a project and have it just work <sup>TM</sup> be it TypeScript, ClojureScript, or plain ol' ES6. If the component can be built with Webpack, then Karma knows how to handle it.

One goal of sapling is to make testing a little less painful. If another framework or library helps with that goal, then it's something we must consider.

This discussion stems from the component [react-flopdown](https://github.com/willowtreeapps/react-flopdown), which is utilizing a different test runner—Jest.

I see some options (not an exhaustive list):

1. **Replace Karma with Jest**
  - Could be an option, but note that some thought will be necessary to get it working with non-ES6 projects (TypeScript, ClojureScript, etc.) It may be _super_ simple, but I haven't looked into Jest + TypeScript setup for several months.
2. **Support Jest _and_ Karma**
  - Freedom for component developers to use what they want at the cost of fragmentation. We already allow components to be written in multiple languages. Do we also want to support multiple test runners?
3. **Learn from Jest & adapt Karma to support things that it can**
  - What things did Jest bring to the table? Where did Karma fall short?

Thoughts? Struggles that you've run into (good and bad) with component testing in the past? Any and all feedback on this is very welcome even if it's short, i.e. "I liked framework X because Y".

**Comment by schreuderjo** [comment]
So I wrote a few tests with Karma before pluggin in Jest.  While the set up to get jest working was a bit of a struggle, once up and running it's a robust testing suite with plenty of documentation.  It's also been used already by other team members of WAT so there are examples in the wild for us to look at already.  Actually that's mostly how I got this set up, by looking at Kevin's configuration.  I see two options for this:

1) Have Jest rolled into sapling

This would require no setup on behalf of the component dev, which might be nice, but then it would restrict customization or tweaking a bit.  Also, it's easy enough should someone NOT wish to use whatever testing suite that comes with sapling to simply rip it out of the package.json and run their own script.  

2) Have a guide on sapling of how to set up Jest testing in each individual component repo at the developers discretion.  Again, now that we have flopdown set up the configuration could be copied over fairly easily.  The downside is it still takes some work, even if it's a little, and that adds friction to the idea of being able to quickly build react components.  

Final thoughts, at least for now it seems like most of these are leaning toward es6 or typescript.  The unknown is how Jest plays with typescript, but it seems like supporting a broad range of language types is good in theory but maybe not necessary in practice.  Perhaps this is shortsighted but can't we cross that bridge when we get there?

**Comment by trevorsenior** [comment]
Thanks for the feedback Jordan. 

> 1) Have Jest rolled into sapling
>
> This would require no setup on behalf of the component dev, which might be nice, but then it would restrict customization or tweaking a bit. Also, it's easy enough should someone NOT wish to use whatever testing suite that comes with sapling to simply rip it out of the package.json and run their own script.

Yep. Totally valid & acceptable for someone to use their own test configuration. I think a no-config default that meets the majority of use cases is valuable be it Karma, Jest, or XYZ.

> 2) Leave it up to the component dev
> 
> We could just leave a guide on sapling's README on how to set up Jest testing in each individual component repo at the developers discretion. Again, now that we have flopdown set up the configuration could be copied over fairly easily. The downside is it still takes some work, even if it's a little, and that adds friction to the idea of being able to quickly build react components.

I think a default test suite that covers the majority of the component test needs is a goal that we want so long as it's overridable.

> Final thoughts, at least for now it seems like most of these are leaning toward es6 or typescript. The unknown is how Jest plays with typescript, but it seems like supporting a broad range of language types is good in theory but maybe not necessary in practice. Perhaps this is shortsighted but can't we cross that bridge when we get there?

With the current state of languages that we would actually use for development this is totally valid. Technically we can force Jest to call out to work with Webpack as necessary and gain all the benefits of it, but yeah. We can totally cross that bridge later.

**Comment by trevorsenior** [comment]
Jest support is coming to sapling to better support existing components projects that wrote their tests using mocks / etc. and wish to switch to react-sapling. See issue 72 for more information.

Closing this for now.

**Comment by kaw2k** [comment]
😂
